### DISCUSSION TITLE
`NSInvalidArgumentException` when using GLFW for Windowing on macOS with Metal

### DISCUSSION THREAD

**Discussion by Emily-TTG** [google/filament#5787] [Q&A] [answered]
Trying to get a basic "Hello World" application running on macOS but am having some trouble with getting started.

With just the following:
```
glfwInit();
glfwWindowHint(GLFW_CLIENT_API, GLFW_NO_API);
GLFWwindow* window = glfwCreateWindow(640, 480, "Hello, Filament!", NULL, NULL);

filament::Engine* engine = filament::Engine::create(filament::Engine::Backend::METAL);
filament::SwapChain* swapChain = engine->createSwapChain(glfwGetCocoaWindow(window));
filament::Renderer* renderer = engine->createRenderer();
filament::View* view = engine->createView();

while(true) {
	if (renderer->beginFrame(swapChain)) {
	    renderer->render(view);
	    renderer->endFrame();
	}
}
```
I am unable to see a window - instead just getting an exception from Cocoa about the GLFW window:
```
Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[GLFWWindow setFramebufferOnly:]: unrecognized selector sent to instance 0x7fdba0707b70'
```
And the backtrace is a bit of a mess so I'll leave it in a pastebin: https://pastebin.com/eT7GYJTx

I am using Filament at the latest trunk as of the time of posting ([efc88fb](https://github.com/google/filament/commit/efc88fbc6f9deb9de2001d078d0cafec11da5351)), using the latest GLFW release available via. macports, macOS Catalina 10.15.7 and compiling with the following using AppleClang 1200.0.32.29:
```
c++ hello_filament.cpp -std=c++17 `pkg-config --cflags --libs glfw3` filament/out/debug/filament/lib/x86_64/* -lobjc -framework CoreFoundation -framework CoreGraphics -framework Cocoa -framework Metal -framework CoreVideo
```

Thanks in advance, and if I missed any info just gimme a holler.

**Comment by bejado** [comment, marked as answer]
You need to ensure that the native window you pass to `createSwapChain` is the correct type. For macOS / Metal, it's a `CAMetalLayer*`. See the [`SwapChain` documentation](https://github.com/google/filament/blob/main/filament/include/filament/SwapChain.h#L42-L50). I'm unfamiliar with GLFW, but maybe [this code sample](https://gist.github.com/gcatlin/987be74e2d58da96093a7598f3fbfb27) will help.

**Reply by Emily-TTG** [reply]
Ah so a surface as opposed to a window for Metal - gotcha, ty!
